tools/crm_mon.c

```diff
@@ -3005,28 +3005,11 @@ reduce_stonith_history(stonith_history_t *history)
                     (hp->state == np->state)) {
                     if ((hp->state == st_done) ||
                         safe_str_eq(hp->delegate, np->delegate)) {
-                        /* replace or purge */
-                        if (hp->completed < np->completed) {
-                            /* purge older hp */
-                            tmp = hp->next;
-                            hp->next = NULL;
-                            stonith_history_free(hp);
-                            hp = tmp;
-                            break;
-                        }
-                        /* damn single linked list */
-                        free(hp->target);
-                        free(hp->action);
-                        free(np->origin);
-                        np->origin = hp->origin;
-                        free(np->delegate);
-                        np->delegate = hp->delegate;
-                        free(np->client);
-                        np->client = hp->client;
-                        np->completed = hp->completed;
-                        tmp = hp;
-                        hp = hp->next;
-                        free(tmp);
+                        /* purge older hp */
+                        tmp = hp->next;
+                        hp->next = NULL;
+                        stonith_history_free(hp);
+                        hp = tmp;
```

Comment:
This looks good to me, but I noticed an interesting consequence. The later block 

        if (np == NULL) {
            tmp = hp->next;
            hp->next = new;
            new = hp;
            hp = tmp;
        }

_prepends_ the entry to "new". That means the previous code (which would later sort the reduced history) would show the newest at the top, but if I'm following correctly, this approach will now have the oldest at the top. I think that's fine and makes sense for the reduced history, though it diverges from the full history display.

Reply:
Hi Ken,

Really?
I did not notice it.
I will check it.

Best Regards,
Hideo Yamauchi.

Reply:
Till now we didn't have sorting for the items shown in the reduced-list.
As they are now sorted I guess we should have them as with the full list (newest first).
Sounds more like just simply chopping off the first item from history, searching for it in the new list and then either dispose or append to the end of new-list. We wouldn't even have to keep a pointer to the last as we could use 'if (np->next == NULL)' to append to this np and break out.

Something like:
```c
static stonith_history_t *
reduce_stonith_history(stonith_history_t *history)
{
    stonith_history_t *new = NULL, *hp, *np, *tmp;

    for (hp = history; hp; ) {
        stonith_history_t *hp_next = hp->next;

        hp->next = NULL;
        if (!new) {
            new = hp;
        } else {
            for (np = new; ; np = np->next) {
                if ((hp->state == st_done) || (hp->state == st_failed)) {
                    /* action not in progress */
                    if (safe_str_eq(hp->target, np->target) &&
                        safe_str_eq(hp->action, np->action) &&
                        (hp->state == np->state) &&
                        ((hp->state == st_done) ||
                         safe_str_eq(hp->delegate, np->delegate)) &&
                        (hp->completed < np->completed)) {
                            /* purge older hp */
                            stonith_history_free(hp);
                            break;
                    }
                }
                if (!np->next) {
                    np->next = hp;
                    break;
                }
            }
        }
        hp = hp_next;
    }   
    return new;
}
```

Reply:
I understand the issue.
Correct the PR and check the operation.

Many thanks!